src/librustc_typeck/check/mod.rs

```diff
@@ -1102,6 +1104,24 @@ fn check_fn<'a, 'gcx, 'tcx>(inherited: &'a Inherited<'a, 'gcx, 'tcx>,
     }
     fcx.demand_suptype(span, ret_ty, actual_return_ty);
 
+    if let Some((id, _)) = *fcx.tcx.sess.entry_fn.borrow() {
+        if id == fn_id {
+            match fcx.sess().entry_type.get() {
+                Some(config::EntryMain) => {
```

Comment:
You'll need to add a [feature gate] to avoid this being insta-stable for types other than `()`.

[feature gate]: https://forge.rust-lang.org/feature-guide.html

Reply:
That should also "fix" the compile-fail tests. Or you could just fix them for real and add `#![feature(termination_trait)]` to them.